### PR TITLE
James 1970 Push security of multi-mailbox search to the back-end

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 1]
```diff
@@ -53,37 +59,66 @@
...
+
+    public static final int DEFAULT_MAXIMUM_LIMIT = 256;
 
     private ElasticSearchIndexer indexer;
     private MessageToElasticSearchJson messageToElasticSearchJson;
     private ElasticSearchListeningMessageSearchIndex testee;
-    private MailboxSession session;
-    private User user;
-    
+
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: session + otherSession to limit renames

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 2]
```diff
@@ -53,37 +59,66 @@
...
+    private User user2;
+
+    private SearchResult user1SearchResult;
+    private SearchResult user2SearchResult;
+
+    private MessageId messageId1;
+    private MessageId messageId2;
+
+    private ElasticSearchSearcher elasticSearchSearcher;
     @Before
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Missing break line

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
  ****************************************************************/
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: The test you added should be on AbstractMessageSearchIndexTest. We want to ensure the different implementation got the same behaviour.

What I can not enforce while simply reading your code.

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java [hunk 4]
```diff
@@ -555,6 +562,7 @@ private Query buildQueryFromMailboxes(ImmutableSet<MailboxId> mailboxIds) {
     private Document createMessageDocument(final MailboxSession session, final MailboxMessage membership) throws MailboxException{
         final Document doc = new Document();
         // TODO: Better handling
+        doc.add(new Field(USERS, session.getUser().getUserName().toUpperCase(Locale.ENGLISH), Store.YES, Index.NOT_ANALYZED));
```
Comment (chibenwa) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: Locale.US please

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 5]
```diff
@@ -56,14 +56,18 @@
 import org.junit.Test;
```
Comment (chibenwa) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: Do it in AbstractMessageSearchIndexTest for all backends please. Thus even less code is needed

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 6]
```diff
@@ -1052,4 +1074,26 @@ public void sortShouldNotDiscardResultWhenSearchingFieldIsIdentical() throws Exc
...
+            .hasSize(1)
+            .containsOnly(m10.getMessageId());
+    }
+
+    @Test
+    public void searchShouldFilterMailboxBelongToMailboxSession() throws Exception {
+        SearchQuery searchQuery = new SearchQuery();
+
+        assertThat(messageSearchIndex.search(session, otherMailbox, searchQuery)).isEmpty();
+        assertThat(messageSearchIndex.search(otherSession, mailbox, searchQuery)).isEmpty();
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: You have 2 asserts. You then need 2 test methods.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 7]
```diff
@@ -1052,4 +1074,26 @@ public void sortShouldNotDiscardResultWhenSearchingFieldIsIdentical() throws Exc
         assertThat(actual).containsOnly(m1.getMessageId(), m2.getMessageId(), m3.getMessageId(), m4.getMessageId(), m5.getMessageId(),
             m6.getMessageId(), m7.getMessageId(), m8.getMessageId(), m9.getMessageId(), mOther.getMessageId(), mailWithAttachment.getMessageId(), mailWithInlinedAttachment.getMessageId());
     }
+
+    @Test
+    public void searchInMultiMailboxShouldReturnMessagesBelongToUserSession() throws Exception {
+        SearchQuery query = new SearchQuery();
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: You have 2 asserts. You then need 2 test methods.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java (hunk 3 above)
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: Ok, I saw the end of the PR. Reset changes in this class.

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java (hunk 5 above)
Comment (chibenwa) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: Ok, I saw the end of the PR. Reset changes in this class.

File: protocols/smtp/dependency-reduced-pom.xml [hunk 8]
```diff
@@ -79,6 +79,12 @@
       <scope>compile</scope>
```
Comment (chibenwa) [inline protocols/smtp/dependency-reduced-pom.xml]: This file should not be commited. Please add it in a .gitignore

File: server/protocols/jmap-integration-testing/pom.xml [hunk 9]
```diff
@@ -34,6 +34,8 @@
 
     <modules>
         <module>jmap-integration-testing-common</module>
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/pom.xml]: We don't want this changeset

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java (hunk 3 above)
Comment (quynhn) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: This one is quite different
- In the AbstractMessageSearchIndexTest, the message is added by manager layer -> then index.
- In the other ones(ElasticSearchListeningMessageSearchIndexTest.java  and LuceneMailboxMessageSearchIndexTest which does not extend for above abstract), the indexing is calling directly via index.add(...). That's why I added those test.

Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java]: It does not make difference to me.

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java (hunk 4 above)
Comment (rouazana) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: why toUpperCase? Generally we store things lowerCase, no?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 10]
```diff
@@ -50,6 +50,11 @@
 
 public abstract class AbstractMessageSearchIndexTest {
 
+    private static final String PRIVATE_NAMESPACE = "#private";
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: You can use directly MailboxConstants.USER_NAMESPACE instead

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 11]
```diff
@@ -1052,4 +1074,26 @@ public void sortShouldNotDiscardResultWhenSearchingFieldIsIdentical() throws Exc
         assertThat(actual).containsOnly(m1.getMessageId(), m2.getMessageId(), m3.getMessageId(), m4.getMessageId(), m5.getMessageId(),
             m6.getMessageId(), m7.getMessageId(), m8.getMessageId(), m9.getMessageId(), mOther.getMessageId(), mailWithAttachment.getMessageId(), mailWithInlinedAttachment.getMessageId());
     }
+
+    @Test
+    public void searchInMultiMailboxShouldReturnMessagesBelongToUserSession() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: s/Belong/Beloging/ (and later)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 12]
```diff
@@ -1082,18 +1082,33 @@ public void searchInMultiMailboxShouldReturnMessagesBelongToUserSession() throws
...
+    @Test
     public void searchShouldFilterMailboxBelongToMailboxSession() throws Exception {
         SearchQuery searchQuery = new SearchQuery();
 
-        assertThat(messageSearchIndex.search(session, otherMailbox, searchQuery)).isEmpty();
+        assertThat(messageSearchIndex.search(otherSession, otherMailbox, searchQuery)).containsOnly(m10.getUid());
+    }
+
+    @Test
+    public void searchShouldEmptyWhenMailboxBelongToAnotherMailboxSession() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: s/Empty/ReturnEmpty/

**Comment by rouazana** [comment]
Maybe we can now remove belongsToNamespaceAndUser method in StoreMailboxManager

**Comment by rouazana** [comment]
I think we'll need to reindex when merging this

**Comment by chibenwa** [comment]
@rouazana Nope, we are not using Lucene 👼

**Comment by quynhn** [comment]
```
Maybe we can now remove belongsToNamespaceAndUser method in StoreMailboxManager
```
No because the search(by mailbox) in StoreMailboxManager does not lean on indexSearch

**Comment by quynhn** [comment]
```
why toUpperCase? Generally we store things lowerCase, no?
```
It's because of the all current indexing using toUpperCase function. Do we need to change them all to "toLowerCase" or just in case of "Users"?

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
```
Expecting:
  <["text/html;charset=iso-8859-1"]>
to contain exactly (and in same order):
  <["application/json; charset=UTF-8"]>
but some elements were not found:
  <["application/json; charset=UTF-8"]>
and others were not expected:
  <["text/html;charset=iso-8859-1"]>

  
Expecting:
  <["text/html;charset=iso-8859-1"]>
to contain exactly (and in same order):
  <["application/json; charset=UTF-8"]>
but some elements were not found:
  <["application/json; charset=UTF-8"]>
and others were not expected:
  <["text/html;charset=iso-8859-1"]>
```

**Comment by chibenwa** [comment]
In which test class ?

**Comment by quynhn** [comment]
test this please

**Comment by quynhn** [comment]
```
org.apache.james.jmap.cassandra.CassandraUserProvisionningConcurrencyTest
07:58:16.607 [ERROR] c.d.d.c.ControlConnection - [Control connection] Cannot connect to any host, scheduling retry in 1000 milliseconds
07:58:17.611 [ERROR] c.d.d.c.ControlConnection - [Control connection] Cannot connect to any host, scheduling retry in 2000 milliseconds

Results :

Failed tests: 
  
Expecting:
 <401>
to be in:
 <[200, 404]>

  
Expecting:
 <401>
to be in:
 <[200, 404]>
```

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**